File: rp_engine/models/session.py
"""Pydantic models for session management."""

from __future__ import annotations

from pydantic import BaseModel


class SessionCreate(BaseModel):
    rp_folder: str
    branch: str = "main"


class SessionResponse(BaseModel):
    id: str
    rp_folder: str
    branch: str
    started_at: str
    ended_at: str | None = None
    metadata: dict | None = None


class TrustChange(BaseModel):
    npc: str
    delta: int
    reason: str


class NewEntity(BaseModel):
    name: str
    type: str
    first_mention_exchange: int | None = None


class SceneProgression(BaseModel):
    first_timestamp: str | None = None
    last_timestamp: str | None = None
    locations_visited: list[str] = []


class PlotThreadStatus(BaseModel):
    thread_id: str
    name: str
    start_counter: int = 0
    end_counter: int = 0


class SessionEndSummary(BaseModel):
    significant_events: list[str] = []
    trust_changes: list[TrustChange] = []
    new_entities: list[NewEntity] = []
    scene_progression: SceneProgression | None = None
    plot_thread_status: list[PlotThreadStatus] = []


class SessionTimelineEntry(BaseModel):
    type: str  # "trust_change" | "event" | "thread_update" | "character_update" | "scene_change" | "continuity_warning"
    exchange_number: int | None = None
    timestamp: str | None = None
    title: str
    detail: dict = {}
    characters: list[str] = []


class SessionTimelineResponse(BaseModel):
    session_id: str
    branch: str
    exchange_range: tuple[int, int]
    entries: list[SessionTimelineEntry]
    entry_counts: dict[str, int] = {}


class UpdateSessionBody(BaseModel):
    metadata: dict


class SessionEndResponse(BaseModel):
    session: SessionResponse
    summary: SessionEndSummary


class SessionSummary(BaseModel):
    session_id: str
    rp_folder: str
    branch: str
    narrative_summary: str
    key_moments: list[dict] = []
    generated_at: str


class Recap(BaseModel):
    rp_folder: str
    branch: str
    session_id: str | None = None
    style: str = "standard"
    recap_text: str
    generated_at: str

File: frontend/src/lib/types/session.ts
export interface SessionCreate {
  rp_folder: string;
  branch?: string;
}
export interface SessionResponse {
  id: string;
  rp_folder: string;
  branch: string;
  started_at: string;
  ended_at: string | null;
  metadata: Record<string, unknown> | null;
}
export interface TrustChange {
  npc: string;
  delta: number;
  reason: string;
}
export interface NewEntity {
  name: string;
  type: string;
  first_mention_exchange: number | null;
}
export interface SceneProgression {
  first_timestamp: string | null;
  last_timestamp: string | null;
  locations_visited: string[];
}
export interface PlotThreadStatus {
  thread_id: string;
  name: string;
  start_counter: number;
  end_counter: number;
}
export interface SessionEndSummary {
  significant_events: string[];
  trust_changes: TrustChange[];
  new_entities: NewEntity[];
  scene_progression: SceneProgression | null;
  plot_thread_status: PlotThreadStatus[];
}
export interface SessionEndResponse {
  session: SessionResponse;
  summary: SessionEndSummary;
}

export interface SessionTimelineEntry {
  type: string;
  exchange_number: number | null;
  timestamp: string | null;
  title: string;
  detail: Record<string, unknown>;
  characters: string[];
}

export interface SessionTimelineResponse {
  session_id: string;
  branch: string;
  exchange_range: [number, number];
  entries: SessionTimelineEntry[];
  entry_counts: Record<string, number>;
}
